src/command/supergraph/compose/do_compose.rs

```diff
@@ -180,52 +197,116 @@ impl Compose {
         f.sync_all()?;
         tracing::debug!("config file written to {}", &yaml_path);
 
-        let federation_version = Self::extract_federation_version(&exe);
+        let federation_version = Self::extract_federation_version(&exe)?;
+        let exact_version = federation_version
+            .get_exact()
+            // This should be impossible to get to because we convert to a FederationVersion a few
+            // lines above and so _should_ have an exact version
+            .ok_or(RoverError::new(anyhow!(
+                "failed to get exact Federation version"
+            )))?;
 
         eprintln!(
             "composing supergraph with Federation {}",
-            &federation_version
+            &federation_version.get_tarball_version()
         );
 
-        let output = Command::new(&exe)
-            .args(["compose", yaml_path.as_ref()])
-            .output()
-            .context("Failed to execute command")?;
-        let stdout = str::from_utf8(&output.stdout)
-            .with_context(|| format!("Could not parse output of `{} compose`", &exe))?;
-
-        match serde_json::from_str::<BuildResult>(stdout) {
-            Ok(build_result) => match build_result {
-                Ok(build_output) => Ok(CompositionOutput {
-                    hints: build_output.hints,
-                    supergraph_sdl: build_output.supergraph_sdl,
-                    federation_version: Some(federation_version.to_string()),
-                }),
-                Err(build_errors) => Err(RoverError::from(RoverClientError::BuildErrors {
-                    source: build_errors,
-                    num_subgraphs,
-                })),
-            },
-            Err(bad_json) => Err(anyhow!("{}", bad_json))
-                .with_context(|| anyhow!("{} compose output: {}", &exe, stdout))
-                .with_context(|| anyhow!("Output from `{} compose` was malformed.", &exe))
-                .map_err(|e| {
-                    let mut error = RoverError::new(e);
-                    error.set_suggestion(RoverErrorSuggestion::SubmitIssue);
-                    error
-                }),
+        // When the `--output` flag is used, we need a supergraph binary version that is at least
+        // v2.9.0. We ignore that flag for composition when we have anything less than that
+        if output_file.is_some() && exact_version.major < 2
```

Comment:
Looks like there's a precedence error here that can lead to the warning being printed when it shouldn't (but no other negative effects). Fixed in https://github.com/apollographql/rover/pull/2100